src/number_theory/legendre_symbol/mul_character.lean

```diff
@@ -0,0 +1,495 @@
+variables (R : Type u) [comm_monoid R]
+-- The target
+variables (R' : Type v) [comm_monoid_with_zero R']
+
+/-- Define a structure for multiplicative characters.
+A multiplicative character from a commutative monoid `R` to a commutative monoid with zero `R'`
+is a homomorphism of (multiplicative) monoids that sends non-units to zero. -/
+structure mul_char extends monoid_hom R R' :=
+(map_nonunit' : ∀ a : R, ¬ is_unit a → to_fun a = 0)
+
+/-- This is the corresponding extension of `monoid_hom_class`. -/
+class mul_char_class (F : Type*) (R R' : out_param $ Type*) [comm_monoid R]
+ [comm_monoid_with_zero R']
+  extends monoid_hom_class F R R' :=
+(map_nonunit : ∀ (χ : F) {a : R} (ha : ¬ is_unit a), χ a = 0)
+
+attribute [simp] mul_char_class.map_nonunit
+
+end defi
+
+section group
+
+namespace mul_char
+
+-- The domain of our multiplicative characters
+variables {R : Type u} [comm_monoid R]
+-- The target
+variables {R' : Type v} [comm_monoid_with_zero R']
+
+instance coe_to_fun : has_coe_to_fun (mul_char R R') (λ _, R → R') :=
+⟨λ χ, χ.to_fun⟩
+
+/-- See note [custom simps projection] -/
+protected def simps.apply (χ : mul_char R R') : R → R' := χ
+initialize_simps_projections mul_char (to_monoid_hom_to_fun → apply, -to_monoid_hom)
+
+section trivial
+
+variables (R R')
+
+/-- The trivial multiplicative character. It takes the value `0` on non-units and
+the value `1` on units. -/
+@[simps]
+noncomputable
+def trivial : mul_char R R' :=
+{ to_fun := by { classical, exact λ x, if is_unit x then 1 else 0 },
+  map_nonunit' := by { intros a ha, simp only [ha, if_false], },
+  map_one' := by simp only [is_unit_one, if_true],
+  map_mul' := by { intros x y,
+                   simp only [is_unit.mul_iff, boole_mul],
+                   split_ifs; tauto, } }
+
+end trivial
+
+
+@[simp]
+lemma coe_coe (χ : mul_char R R') : (χ.to_monoid_hom : R → R') = χ := rfl
+
+@[simp]
+lemma to_fun_eq_coe (χ : mul_char R R') : χ.to_fun = χ := rfl
+
+@[simp]
+lemma coe_mk (f : R →* R') (hf) : (mul_char.mk f hf : R → R') = f := rfl
+
+/-- Extensionality. See `ext` below for the version that will actually be used. -/
+lemma ext' {χ χ' : mul_char R R'} (h : ∀ a, χ a = χ' a) : χ = χ' :=
+begin
+  cases χ,
+  cases χ',
+  congr,
+  exact monoid_hom.ext h,
+end
+
+instance : mul_char_class (mul_char R R') R R' :=
+{ coe := λ χ, χ.to_monoid_hom.to_fun,
+  coe_injective' := λ f g h, ext' (λ a, congr_fun h a),
+  map_mul := λ χ, χ.map_mul',
+  map_one := λ χ, χ.map_one',
+  map_nonunit := λ χ, χ.map_nonunit', }
+
+lemma map_nonunit (χ : mul_char R R') {a : R} (ha : ¬ is_unit a) : χ a = 0 :=
+χ.map_nonunit' a ha
+
+/-- Extensionality. Since `mul_char`s always take the value zero on non-units, it is sufficient
+to compare the values on units. -/
+@[ext]
+lemma ext {χ χ' : mul_char R R'} (h : ∀ a : Rˣ, χ a = χ' a) : χ = χ' :=
+begin
+  apply ext',
+  intro a,
+  by_cases ha : is_unit a,
+  { exact h ha.unit, },
+  { rw [map_nonunit χ ha, map_nonunit χ' ha], },
+end
```

Comment:
This will help to prove `is_nontrivial_iff`:
```suggestion
end

lemma ext_iff {χ χ' : mul_char R R'} : χ = χ' ↔ ∀ a : Rˣ, χ a = χ' a :=
⟨by { rintro rfl a, refl }, ext⟩
```

Reply:
OK.